### ISSUE TITLE
ESLint v5 Wish List

### ISSUE THREAD

**Issue by btmills** [eslint/eslint#8972]
In the spirit of [previous wish lists](https://github.com/eslint/eslint/issues/2203), now that v4 has shipped, what big changes do we want to see in v5? We can use this issue for brainstorming and discussing ideas. For the changes we want to tackle for v5, we'll then create issues in the [v5.0.0 milestone](https://github.com/eslint/eslint/milestone/18).

**Comment by platinumazure** [comment]
I'd like to see if we can get rule messages into metadata (see #6740-- concept has been accepted and is on core roadmap, but hasn't been important enough to get onto a major version milestone yet).

**Comment by platinumazure** [comment]
Also, implementing #7443 (warnings on deprecated rules) would be awesome.

**Comment by not-an-aardvark** [comment]
* Linting files in parallel (https://github.com/eslint/eslint/issues/3565) would benefit a lot of people.
* Allowing parsers to hook into scope analysis (https://github.com/eslint/eslint/pull/8755) would make it much easier for parsers to support custom syntax.
* Allowing autofixes in preprocessors (https://github.com/eslint/eslint/issues/7510) would make them much more useful -- adding a preprocessor currently imposes a high cost on the user.
* Eliminating the behavior change between local and global installs (https://github.com/eslint/eslint/issues/6732). In my opinion, the distinction has caused much more confusion than it's worth.
* Dropping support for Node 4, as it's no longer in LTS. This would allow us to use packages that support Node 6+, and we could also use more ES6 in our codebase.

Stretch goals (on my wishlist, but not sure they'll get done by v5.x):

* Creating a better way to include plugins in shareable configs (https://github.com/eslint/eslint/issues/3458)
* Creating a better API for autofixes that modify the AST. With our existing text-based autofix API, it's extremely difficult to write a correct autofixer that modifies the AST, because there are so many cases to handle. You have to make sure that the fix (a) doesn't affect precedence, (b) doesn't combine two tokens into one as a result of no spacing, (c) doesn't combine two statements into one as a result of ASI, (d) doesn't cause some code to become a comment, (e) doesn't put illegal nodes such as function expressions at the beginning of a statement without parenthesizing them, (f) doesn't cause a statement to become a directive, (g) doesn't create mismatched parentheses if a fixed node is parenthesized, (h) doesn't remove a node from a location where it's required, ... etc.

    I'm imagining that we can have create like the Babel plugin API to make modifications to the AST and handle formatting automatically.

**Comment by Aladdin-ADD** [comment]
- Per-rule autofixing(#8018): this is useful to some editor.

**Comment by ilyavolodin** [comment]
All of the things mentioned above plus a few additional things:
1. Most of the editor integrations don't use our APIs, and instead build wrappers around CLI (at least that's what I've heard). I would like to understand what's missing from the API and how we can improve it to make it easier to integrate with ESLint.
2. I think we need to spend some cycles on performance enhancements. We've neglected that aspect for a while, and to be honest ESLint is madly slow on large repositories. We can start by restoring our NPM perf script.
3. Now that glob configs are in, I think we need to rethink preprocessors and how they are registered.
4. I think maybe it's time we start thinking about breaking ESLint into smaller packages. @nzakas had a (very) long term vision of breaking ESLint into reusable parser/linter that can be used as a base for other linters. I think would could think about even smaller chunks that can be broken off. Like config management and loading.

**Comment by fgarcia** [comment]
I am the the OP #8949 and as mentioned in the comments my request was not about per-rule autofixing

My concern is that ESLint serves two purposes

  1. Potential errors
  2. Code style / beautification

I wanted an easy way to run ESLint in one of those two modes. I assume that all `--fix` rules are about code beautification. In my workflow I want (1) before commit and (2) while editing.

Obviously I fixed this by disabling one by one every `--fix` rule. I try to use ESLint to beautify my code (2), not to force me to beautify my own code. 

Personally I even believe that per-rule autofixing will add more complexity than needed. Mostly because I have two workflows where having `--fix` rules is all or nothing. 

After reading the comments I understood that my workflow / need is quite rare, so I did not want to discuss the topic further if it has little value for others. However if that issue is added to a wish list, I would insist again that it is not about "per-rule autofixing", I am against that. My wish is about having an easy way to split two tasks: error detection vs beautification

**Comment by platinumazure** [comment]
@ilyavolodin

> Most of the editor integrations don't use our APIs, and instead build wrappers around CLI (at least that's what I've heard). I would like to understand what's missing from the API and how we can improve it to make it easier to integrate with ESLint.

How much of that, in your view, is actually lack of API surface? Personally, I think it more likely that a lot of editor environments make it easier to invoke shell scripts or commands rather than invoking a Node runtime.

**Comment by ilyavolodin** [comment]
@platinumazure I'm not really sure, to be honest. I know that both Atom and VSC are capable of using our Node API, but neither of does. I think mostly this is due to performance, but, again, don't have enough information.

**Comment by not-an-aardvark** [comment]
Actually, the Atom integration does use the Node API as of a few months ago (https://github.com/AtomLinter/linter-eslint/pull/873)

**Comment by not-an-aardvark** [comment]
> I would like to understand what's missing from the API and how we can improve it to make it easier to integrate with ESLint.

I think we should also improve our documentation. There are a lot of undocumented methods methods in the API, which might discourage people from using it because they might not know something exists.

**Comment by j-f1** [comment]
I would love to see the ESLint and Babel ASTs merged. This would make ESLint better because new parser features only have to get implemented in one parser, reducing the ESLint team’s workload. The ASTs are mostly compatible, but Babylon’s has a few improvements IMO:

* Different literals have different node types. For example, `2` is a `NumericLiteral`, while `"foo"` is a `StringLiteral`
* `foo () {}` in `{ foo () {} }` and `foo: function () {}` in `{ foo: function () {} }` are different types of node. 
* New JS feature proposals are supported, which means that people will be able to write lint rules for them.

Also, `babel-types` is awesome.

A quick look through the AST shows that many rules would probably continue to work fine, and the ones that don’t should be able to be fixed fairly easily.

<details><summary>Changed nodes</summary>

- [`RegExpLiteral`](https://github.com/babel/babylon/blob/master/ast/spec.md#regexpliteral ) ([ESTree](https://github.com/estree/estree/blob/master/es5.md#regexpliteral)): The `pattern` and `flags` properties are moved to the root of the node, instead of inside a `regex` property. This could be made backwards-compatible.
- [`NullLiteral`](https://github.com/babel/babylon/blob/master/ast/spec.md#nullliteral), [`StringLiteral`](https://github.com/babel/babylon/blob/master/ast/spec.md#stringliteral), [`BooleanLiteral`](https://github.com/babel/babylon/blob/master/ast/spec.md#booleanliteral), [`NumericLiteral`](https://github.com/babel/babylon/blob/master/ast/spec.md#numericliteral): Split out from `Literal`. Somewhat backwards compatible — the `Literal` listener can be called with the old node, while the `*Literal` types can be called with the new node.
- [`Program`](https://github.com/babel/babylon/blob/master/ast/spec.md#programs) ([ESTree](https://github.com/babel/babylon/blob/master/ast/spec.md#programs)), [`BlockStatement`](https://github.com/babel/babylon/blob/master/ast/spec.md#blockstatement) ([ESTree](https://github.com/estree/estree/blob/master/es5.md#blockstatement)): New `directives` key
- [`ForOfStatement`](https://github.com/babel/babylon/blob/master/ast/spec.md#forofstatement) ([ESTree](https://github.com/estree/estree/blob/master/es2015.md#forofstatement)): New `await` key.
- [`AwaitExpression`](https://github.com/babel/babylon/blob/master/ast/spec.md#awaitexpression) ([ESTree](https://github.com/estree/estree/blob/master/es2017.md#awaitexpression)): `argument` becomes optional 😕
- [`ObjectExpression`](https://github.com/babel/babylon/blob/master/ast/spec.md#objectexpression) ([ESTree](https://github.com/estree/estree/blob/master/es5.md#objectexpression)): `properties` can have `ObjectProperty | ObjectMethod | SpreadElement`
- [`ObjectProperty`](https://github.com/babel/babylon/blob/master/ast/spec.md#objectproperty), [`ObjectMethod`](https://github.com/babel/babylon/blob/master/ast/spec.md#objectmethod): Split from ESTree’s [`Property`](https://github.com/estree/estree/blob/master/es5.md#property) ([ES6 changes](https://github.com/estree/estree/blob/master/es2015.md#expressions)). They have `decorators`, and `kind` is slightly different.
- [`MemberExpression`](https://github.com/babel/babylon/blob/master/ast/spec.md#memberexpression) ([ESTree](https://github.com/estree/estree/blob/master/es5.md#memberexpression)), [`CallExpression`](https://github.com/babel/babylon/blob/master/ast/spec.md#callexpression) ([ESTree](https://github.com/estree/estree/blob/master/es5.md#callexpression)), [`NewExpression`](https://github.com/babel/babylon/blob/master/ast/spec.md#newexpression) ([ESTree](https://github.com/estree/estree/blob/master/es5.md#newexpression)): New `optional` property
- [`CallExpression`](https://github.com/babel/babylon/blob/master/ast/spec.md#callexpression) ([ESTree](https://github.com/estree/estree/blob/master/es5.md#callexpression)): `callee` can be an `Import` node
- [`TemplateElement`](https://github.com/babel/babylon/blob/master/ast/spec.md#templateelement) ([ESTree](https://github.com/estree/estree/blob/master/es2015.md#templateelement)): `value.cooked` can be `null`
- [`AssignmentProperty`](https://github.com/babel/babylon/blob/master/ast/spec.md#objectpattern) ([ESTree](https://github.com/estree/estree/blob/master/es2015.md#objectpattern)): `kind` is gone, the node now has a `type` of `ObjectProperty` instead of `Property` 
- [`ObjectPattern`](https://github.com/babel/babylon/blob/master/ast/spec.md#objectpattern) ([ESTree](https://github.com/estree/estree/blob/master/es2015.md#objectpattern)): `properties` can include [`RestElement`](https://github.com/babel/babylon/blob/master/ast/spec.md#restelement)s
- [`Class`](https://github.com/babel/babylon/blob/master/ast/spec.md#classes) ([ESTree](https://github.com/estree/estree/blob/master/es2015.md#classes)): New `decorators` property
- [`ClassBody`](https://github.com/babel/babylon/blob/master/ast/spec.md#classbody) ([ESTree](https://github.com/estree/estree/blob/master/es2015.md#classbody)): Body can contain `ClassProperty | ClassPrivateProperty`
- [`ClassMethod`](https://github.com/babel/babylon/blob/master/ast/spec.md#classmethod) ([ESTree: `MethodDefinition`](https://github.com/estree/estree/blob/master/es2015.md#methoddefinition)): *is* a `Function`, instead of having a `FunctionExpression` in `value`; new `decorators` property

New:
- [`Decorator`](https://github.com/babel/babylon/blob/master/ast/spec.md#decorator)
- [`Directive`](https://github.com/babel/babylon/blob/master/ast/spec.md#directive)
- [`DirectiveLiteral`](https://github.com/babel/babylon/blob/master/ast/spec.md#directiveliteral)
- [`Import`](https://github.com/babel/babylon/blob/master/ast/spec.md#import)
- [`BindExpression`](https://github.com/babel/babylon/blob/master/ast/spec.md#bindexpression)
- [`DoExpression`](https://github.com/babel/babylon/blob/master/ast/spec.md#doexpression)
- [`ClassProperty`](https://github.com/babel/babylon/blob/master/ast/spec.md#classproperty)
- [`ClassPrivateProperty`](https://github.com/babel/babylon/blob/master/ast/spec.md#classprivateproperty)

**Comment by nzakas** [comment]
Babel unfortunately uses a nonstandard AST. We use the ESTree format, and Babel early on started out with that format before going out on its own. I'd rather support the standard that is implemented by many projects (Esprima, Acorn, escope, estraverse, and of course ESLint and Espree) as opposed to adopting something that only one tool used.

**Comment by vanduynslagerp** [comment]
ESLint has become the standard for Javascript linting and it is increasingly popular. So much that a very large number of developer/company/organization created their own shareable config.
A [search](https://www.npmjs.com/search?q=eslint-config-*&page=1&ranking=optimal) for `eslint-config-*` on NPM returns 3047 results as today !

So I'd like to emphasis the importance to find a solution to #3458. Also it's the issue with currently the most comment/reactions, so it's seems ESLint users find it important.

This problem is really not easy to solve, so the debate on this issue has been going on for 2 years. But is seems there might be a consensus around @not-an-aardvark's solution proposed [here](https://github.com/eslint/eslint/issues/3458#issuecomment-266594020).

So v5 might be a good time to tackle this issue.

**Comment by j-f1** [comment]
> 2. Similarly, id like to move the CommonJS/NodeJS specific rules into a plugin. As the world increasingly moves towards async functions and ES modules, these will become less useful to the general population of devs.

[`eslint-plugin-node`](https://github.com/mysticatea/eslint-plugin-node) might be happy to accept those rules.

**Comment by nzakas** [comment]
@j-f1 that's exactly what I was thinking. Good to know we are on the same page. 

Another idea: I'd like to update eslint:recommended with recommendations for post-ES5 code, as I think there are emerging best practices that would be useful to get out there. We can create eslint:recommended-legacy if we want to provide recommendations for ES5 only (though we should discuss if that's necessary).

**Comment by Aladdin-ADD** [comment]
closing as eslint v5 has been released.
